libcontainer/cgroups/systemd/v2.go

```diff
@@ -288,6 +290,36 @@ func (m *unifiedManager) Apply(pid int) error {
 	if err := fs2.CreateCgroupPath(m.path, m.cgroups); err != nil {
 		return err
 	}
+
+	if c.OwnerUID != nil {
+		// The kernel exposes a list of files that should be chowned to
+		// the delegate uid in /sys/kernel/cgroup/delegate.  If the
+		// file is not present (Linux < 4.15), use the initial
+		// values mentioned in cgroups(7).
+		filesToChown := []string{"."} // the directory itself must be chowned
+		cgroupDelegateFile := "/sys/kernel/cgroup/delegate"
+		f, err := os.Open(cgroupDelegateFile)
+		if err == nil {
+			defer f.Close()
+			scanner := bufio.NewScanner(f)
+			for scanner.Scan() {
+				filesToChown = append(filesToChown, scanner.Text())
+			}
+			if err := scanner.Err(); err != nil {
+				return fmt.Errorf("error reading %s: %w", cgroupDelegateFile, err)
+			}
+		} else {
+			filesToChown = append(filesToChown, "cgroup.procs", "cgroup.subtree_control", "cgroup.threads")
+		}
+
+		for _, v := range filesToChown {
```

Comment:
Perhaps it's better to separate out the code that finds files to be chowned to a separate `filesToChown()` function, and just call it from below, e.g.

```go
for _, v := range filesToChown() {
...
}
```

This should increase readability and decrease Apply() complexity.

The only thing is, to simplify, you would not return any errors from filesToChown(), instead maybe logging it as warnings and falling back to built-in list.

Reply:
I extracted the behaviour to another function.  But I kept the error return.  I feel that `/sys/kernel/cgroup/delegate` existing but being unreadable is something we should fail hard on, because its values may (and currently do) differ from the original list.

Reply:
Thanks, this looks better and more reviewable!